File: Data Structures & Algorithms/course-schedule/submission-3.js
class Solution {
    /**
     * @param {number} numCourses
     * @param {number[][]} prerequisites
     * @return {boolean}
     */
    canFinish(numCourses, prerequisites) {
        const adj = Array.from({ length: numCourses }, () => []);
        const indegree = new Array(numCourses).fill(0);

        for(const [c, p] of prerequisites) {
            adj[p].push(c);
            indegree[c]++;
        }

        const q = []
        for(let i = 0; i < numCourses; ++i) if(indegree[i] === 0) q.push(i);

        let cnt = 0;
        let head = 0;
        while(head < q.length) {
            const u = q[head++];
            cnt++;

            for(const v of adj[u]) {
                if(--indegree[v] === 0) q.push(v);
            }
        }

        return cnt == numCourses;
    }
}

File: Data Structures & Algorithms/course-schedule-ii/submission-3.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(numCourses)]
        indegree = [0] * numCourses

        for c, p in prerequisites:
            adj[p].append(c)
            indegree[c] += 1
        
        queue = collections.deque([i for i in range(numCourses) if indegree[i] == 0])

        ans = []
        while queue:
            u = queue.popleft()
            ans.append(u)

            for v in adj[u]:
                indegree[v] -= 1
                if indegree[v] == 0:
                    queue.append(v)
        
        return ans if len(ans) == numCourses else []